File: src/main/devlog.ts
/**
 * 开发过程档案（devlog）。
 *
 * 每做一轮改动就是一个「轮次」，落在 devlog/rounds/<编号>-<时间戳>/ 下，
 * 内含：用户标注原文、逐条标注的截图裁片、当轮改动说明、代码差异。
 * 目的是让整个项目的演进过程可回溯、可展示。
 *
 * 标注由界面上的「标注工具」产生，导出时主进程负责截图与落盘。
 */
import { existsSync } from 'node:fs'
import { mkdir, readdir, readFile, writeFile } from 'node:fs/promises'
import { join } from 'node:path'
import { app, type BrowserWindow } from 'electron'

export interface AnnotationRect {
  x: number
  y: number
  width: number
  height: number
}

export interface AnnotationTarget {
  selector: string
  component: string | null
  tag: string
  classes: string
  text: string
  parentChain: string[]
  styles: Record<string, string>
  attributes: Record<string, string>
}

export interface Annotation {
  id: string
  index: number
  comment: string
  category: 'style' | 'layout' | 'content' | 'behavior' | 'bug' | 'other'
  severity: 'must' | 'should' | 'nice'
  kind: 'element' | 'region'
  createdAt: number
  rect: AnnotationRect
  viewport: { width: number; height: number }
  target: AnnotationTarget
}

export interface ExportPayload {
  annotations: Annotation[]
  /** 本轮的整体说明，可选 */
  note: string
  /** 标注时所在界面（gallery / lightbox / crawl / settings …） */
  view: string
  devicePixelRatio: number
  appVersion: string
}

export interface RoundInfo {
  id: string
  dir: string
  createdAt: string
  annotationCount: number
  hasScreenshot: boolean
}

const CATEGORY_LABEL: Record<Annotation['category'], string> = {
  style: '视觉样式',
  layout: '布局结构',
  content: '文案内容',
  behavior: '交互行为',
  bug: '缺陷',
  other: '其它'
}

const SEVERITY_LABEL: Record<Annotation['severity'], string> = {
  must: '必须改',
  should: '建议改',
  nice: '锦上添花'
}

export class DevlogStore {
  constructor(private readonly root: string) {}

  /**
   * 档案根目录的定位顺序：
   *   1. GUGU_WORKSPACE 环境变量（显式指定）
   *   2. 应用目录下的 devlog/（开发态就是仓库根）
   *   3. 用户数据目录（打包后 asar 不可写时的兜底，保证导出永远能成功）
   */
  static resolveRoot(): string {
    const fromEnv = process.env.GUGU_WORKSPACE
    if (fromEnv) return fromEnv
    const appPath = app.getAppPath()
    if (!appPath.includes('app.asar') && existsSync(join(appPath, 'devlog'))) return appPath
    return join(app.getPath('userData'), 'devlog-workspace')
  }

  get roundsDir(): string {
    return join(this.root, 'devlog', 'rounds')
  }

  async listRounds(): Promise<RoundInfo[]> {
    if (!existsSync(this.roundsDir)) return []
    const entries = await readdir(this.roundsDir, { withFileTypes: true })
    const rounds: RoundInfo[] = []
    for (const entry of entries) {
      if (!entry.isDirectory()) continue
      const dir = join(this.roundsDir, entry.name)
      let annotationCount = 0
      try {
        const raw = await readFile(join(dir, 'annotations.json'), 'utf8')
        const parsed = JSON.parse(raw) as { annotations?: unknown[] }
        annotationCount = parsed.annotations?.length ?? 0
      } catch {
        /* 没有标注文件也正常（纯代码轮次） */
      }
      rounds.push({
        id: entry.name,
        dir,
        createdAt: entry.name.match(/^(\d{4})(\d{2})(\d{2})-(\d{2})(\d{2})/)?.[0] ?? entry.name,
        annotationCount,
        hasScreenshot: existsSync(join(dir, 'screenshots', '00-full.png'))
      })
    }
    return rounds.sort((a, b) => a.id.localeCompare(b.id))
  }

  /** 下一个轮次目录：<4 位序号>-<yyyyMMdd-HHmm> */
  async nextRoundDir(): Promise<{ id: string; dir: string }> {
    const rounds = await this.listRounds()
    const nextNumber = rounds.length + 1
    const id = `${String(nextNumber).padStart(4, '0')}-${stamp()}`
    return { id, dir: join(this.roundsDir, id) }
  }

  /**
   * 导出一次标注：整页截图 + 每条标注的裁片 + JSON + Markdown。
   * 截图裁切要注意 devicePixelRatio：capturePage 给的是物理像素，标注框是 CSS 像素。
   */
  async exportAnnotations(
    payload: ExportPayload,
    win: BrowserWindow | null
  ): Promise<{ roundId: string; roundDir: string; files: string[] }> {
    const { id, dir } = await this.nextRoundDir()
    const shotsDir = join(dir, 'screenshots')
    await mkdir(shotsDir, { recursive: true })

    const files: string[] = []

    // 1) 整页截图
    let fullImage: Electron.NativeImage | null = null
    if (win && !win.isDestroyed()) {
      win.webContents.invalidate()
      await sleep(320)
      fullImage = await win.webContents.capturePage()
      await writeFile(join(shotsDir, '00-full.png'), fullImage.toPNG())
      files.push('screenshots/00-full.png')
    }

    // 2) 每条标注裁片
    const dpr = payload.devicePixelRatio || 1
    for (const anno of payload.annotations) {
      if (!fullImage) break
      const name = `${String(anno.index).padStart(3, '0')}-${slugify(anno.comment)}.png`
      try {
        const size = fullImage.getSize()
        const x = clamp(Math.round(anno.rect.x * dpr) - 12, 0, size.width)
        const y = clamp(Math.round(anno.rect.y * dpr) - 12, 0, size.height)
        const width = clamp(Math.round(anno.rect.width * dpr) + 24, 1, size.width - x)
        const height = clamp(Math.round(anno.rect.height * dpr) + 24, 1, size.height - y)
        const crop = fullImage.crop({ x, y, width, height })
        if (!crop.isEmpty()) {
          await writeFile(join(shotsDir, name), crop.toPNG())
          files.push(`screenshots/${name}`)
        }
      } catch {
        /* 裁切失败不影响主体导出 */
      }
    }

    // 3) 原始数据
    const record = {
      roundId: id,
      exportedAt: new Date().toISOString(),
      view: payload.view,
      note: payload.note,
      viewport: payload.annotations[0]?.viewport ?? null,
      devicePixelRatio: dpr,
      appVersion: payload.appVersion,
      annotations: payload.annotations
    }
    await writeFile(join(dir, 'annotations.json'), JSON.stringify(record, null, 2), 'utf8')
    files.push('annotations.json')

    // 4) 人能读的 Markdown
    await writeFile(join(dir, 'annotations.md'), renderMarkdown(record), 'utf8')
    files.push('annotations.md')

    // 5) 轮次说明骨架，供后续填写
    if (!existsSync(join(dir, 'README.md'))) {
      await writeFile(join(dir, 'README.md'), renderRoundReadme(record), 'utf8')
      files.push('README.md')
    }

    return { roundId: id, roundDir: dir, files }
  }
}

/* ------------------------------------------------------------------ 渲染 */

interface ExportRecord {
  roundId: string
  exportedAt: string
  view: string
  note: string
  viewport: { width: number; height: number } | null
  devicePixelRatio: number
  appVersion: string
  annotations: Annotation[]
}

function renderRoundReadme(record: ExportRecord): string {
  const viewport = record.viewport ? `${record.viewport.width}×${record.viewport.height}` : '未知'
  return `# 轮次 ${record.roundId}

> 由界面上的「标注工具」自动生成，请在完成本轮改动后补全下面的小节。

## 背景

<!-- 这一轮要解决什么问题、由谁提出 -->

${record.note ? record.note : '（待补充）'}

## 反馈标注

共 ${record.annotations.length} 条，详见 [\`annotations.md\`](annotations.md) 与 [\`annotations.json\`](annotations.json)。

导出时视口 ${viewport}，界面 ${record.view}，应用版本 ${record.appVersion}。

## 改动

<!-- 逐条说明做了什么，以及为什么这么做（含取舍） -->

## 验证

<!-- 跑了哪些测试、截图证据、以及没能验证到的部分 -->

## 遗留

<!-- 本轮没做、留待下轮的事项 -->
`
}

function renderMarkdown(record: ExportRecord): string {
  const lines: string[] = []
  lines.push(`# 界面标注 · 轮次 ${record.roundId}`)
  lines.push('')
  lines.push(`- 导出时间：${record.exportedAt}`)
  lines.push(`- 所在界面：${record.view}`)
  if (record.viewport) lines.push(`- 视口：${record.viewport.width}×${record.viewport.height}（DPR ${record.devicePixelRatio}）`)
  lines.push(`- 应用版本：${record.appVersion}`)
  lines.push(`- 标注条数：${record.annotations.length}`)
  if (record.note) {
    lines.push('')
    lines.push('## 总体说明')
    lines.push('')
    lines.push(record.note)
  }
  lines.push('')
  lines.push('## 逐条标注')
  lines.push('')

  for (const a of record.annotations) {
    const t = a.target
    lines.push(`### ${a.index}. ${firstLine(a.comment) || '(无标题)'}`)
    lines.push('')
    lines.push(
      `- **类别**：${CATEGORY_LABEL[a.category]} ｜ **优先级**：${SEVERITY_LABEL[a.severity]} ｜ **方式**：${
        a.kind === 'element' ? '点选元素' : '框选区域'
      }`
    )
    if (t.component) lines.push(`- **组件**：\`${t.component}\``)
    lines.push(`- **选择器**：\`${t.selector}\``)
    if (t.parentChain.length > 0) lines.push(`- **父级链**：${t.parentChain.map((p) => `\`${p}\``).join(' ← ')}`)
    if (t.text) lines.push(`- **元素文本**：${truncate(t.text.replace(/\s+/g, ' '), 120)}`)
    lines.push(
      `- **位置尺寸**：x=${Math.round(a.rect.x)} y=${Math.round(a.rect.y)} ${Math.round(a.rect.width)}×${Math.round(
        a.rect.height
      )}`
    )
    const styleEntries = Object.entries(t.styles)
    if (styleEntries.length > 0) {
      lines.push(`- **关键样式**：${styleEntries.map(([k, v]) => `${k}: ${v}`).join('; ')}`)
    }
    const attrEntries = Object.entries(t.attributes)
    if (attrEntries.length > 0) {
      lines.push(`- **属性**：${attrEntries.map(([k, v]) => `${k}="${truncate(v, 40)}"`).join(' ')}`)
    }
    lines.push('')
    lines.push('> ' + a.comment.split('\n').join('\n> '))
    lines.push('')
  }
  return lines.join('\n')
}

/* ------------------------------------------------------------------ 工具 */

function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value))
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

function firstLine(text: string): string {
  return (text.split('\n')[0] ?? '').trim().slice(0, 60)
}

function truncate(text: string, max: number): string {
  return text.length <= max ? text : `${text.slice(0, max)}…`
}

function slugify(text: string): string {
  const cleaned = firstLine(text)
    // 半角与全角标点都可能出现在文件名里，统一清掉；中文本身保留
    .replace(/[\\/:*?"<>|：；？！，。、（）【】《》""''…—]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
  return cleaned.slice(0, 28) || 'annotation'
}

function stamp(): string {
  const d = new Date()
  const pad = (n: number): string => String(n).padStart(2, '0')
  return `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}-${pad(d.getHours())}${pad(d.getMinutes())}`
}

File: src-tauri/src/plugins/devlog.rs
//! devlog 插件（Rust 侧）：轮次目录管理 + 标注导出。
//!
//! 与 plugins/devlog/main/{store,render}.ts 的行为对齐：
//! 目录布局、文件顺序、裁切留白、Markdown 结构都保持原样，
//! 这样 devlog/rounds 里已经归档的轮次与 bin/round.mjs 的工作流都不用改。

use std::fs;
use std::path::Path;
use std::time::Duration;

use chrono::{Local, SecondsFormat, Utc};
use serde::Deserialize;
use serde_json::{json, Map, Value as Json};
use tauri::Manager;

use crate::shot;

#[derive(Deserialize)]
struct Rect {
    x: f64,
    y: f64,
    width: f64,
    height: f64,
}

#[derive(Deserialize, Default)]
#[serde(rename_all = "camelCase")]
struct Target {
    #[serde(default)]
    selector: String,
    #[serde(default)]
    component: Option<String>,
    #[serde(default)]
    tag: String,
    #[serde(default)]
    classes: String,
    #[serde(default)]
    text: String,
    #[serde(default)]
    parent_chain: Vec<String>,
    #[serde(default)]
    styles: Map<String, Json>,
    #[serde(default)]
    attributes: Map<String, Json>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Annotation {
    index: i64,
    comment: String,
    category: String,
    severity: String,
    kind: String,
    rect: Rect,
    #[serde(default)]
    viewport: Option<Json>,
    target: Target,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ExportPayload {
    annotations: Vec<Annotation>,
    #[serde(default)]
    note: String,
    #[serde(default)]
    view: String,
    #[serde(default = "default_dpr")]
    device_pixel_ratio: f64,
    #[serde(default)]
    app_version: String,
}

fn default_dpr() -> f64 {
    1.0
}

/// 渲染 Markdown 与 Readme 时需要的全部信息
struct ExportRecord<'a> {
    round_id: &'a str,
    exported_at: &'a str,
    view: &'a str,
    note: &'a str,
    viewport: Option<(f64, f64)>,
    dpr: f64,
    app_version: &'a str,
    annotations: &'a [Annotation],
}

/* ------------------------------------------------------------------ 轮次 */

pub fn list_rounds(workspace: &Path) -> Result<Json, String> {
    let rounds_dir = workspace.join("devlog").join("rounds");
    if !rounds_dir.is_dir() {
        return Ok(json!([]));
    }

    let mut rounds: Vec<Json> = Vec::new();
    let entries = fs::read_dir(&rounds_dir).map_err(|e| format!("读取轮次目录失败: {e}"))?;
    for entry in entries.flatten() {
        let dir = entry.path();
        if !dir.is_dir() {
            continue;
        }
        let name = entry.file_name().to_string_lossy().to_string();
        let annotation_count = fs::read_to_string(dir.join("annotations.json"))
            .ok()
            .and_then(|raw| serde_json::from_str::<Json>(&raw).ok())
            .and_then(|v| v.get("annotations")?.as_array().map(|list| list.len()))
            .unwrap_or(0);
        rounds.push(json!({
            "id": name,
            "dir": dir.to_string_lossy(),
            "createdAt": created_at(&name),
            "annotationCount": annotation_count,
            "hasScreenshot": dir.join("screenshots").join("00-full.png").exists()
        }));
    }
    rounds.sort_by(|a, b| {
        a.get("id")
            .and_then(Json::as_str)
            .unwrap_or("")
            .cmp(b.get("id").and_then(Json::as_str).unwrap_or(""))
    });
    Ok(Json::Array(rounds))
}

/// 目录名形如 0014-20260916-1802 时取出 20260916-1802；否则原样返回（旧版判据是 /^\d{8}-\d{4}/）
fn created_at(name: &str) -> String {
    let bytes = name.as_bytes();
    let digits = |start: usize, len: usize| {
        bytes.len() >= start + len && bytes[start..start + len].iter().all(u8::is_ascii_digit)
    };
    if digits(0, 8) && bytes.get(8) == Some(&b'-') && digits(9, 4) {
        return name[..13].to_string();
    }
    name.to_string()
}

/* ------------------------------------------------------------ 标注导出 */

pub fn export_annotations(
    app: &tauri::AppHandle,
    workspace: &Path,
    payload: Json,
) -> Result<Json, String> {
    let data: ExportPayload =
        serde_json::from_value(payload.clone()).map_err(|e| format!("导出参数不合法：{e}"))?;
    // 落盘的 annotations.json 要保留原始标注对象（含 pageRect / scroll / target），不能过结构体再序列化
    let raw_annotations = payload.get("annotations").cloned().unwrap_or_else(|| json!([]));

    let rounds_dir = workspace.join("devlog").join("rounds");
    let round_id = format!("{:04}-{}", count_rounds(&rounds_dir) + 1, stamp());
    let dir = rounds_dir.join(&round_id);
    let shots_dir = dir.join("screenshots");
    fs::create_dir_all(&shots_dir).map_err(|e| format!("创建轮次目录失败: {e}"))?;

    let mut files: Vec<String> = Vec::new();
    let dpr = if data.device_pixel_ratio > 0.0 {
        data.device_pixel_ratio
    } else {
        1.0
    };

    if let Some(full) = capture_full(app) {
        let full_path = shots_dir.join("00-full.png");
        full.save_with_format(&full_path, image::ImageFormat::Png)
            .map_err(|e| format!("写入整页截图失败: {e}"))?;
        files.push("screenshots/00-full.png".to_string());

        let (image_width, image_height) = full.dimensions();
        for annotation in &data.annotations {
            let name = format!("{:03}-{}.png", annotation.index, slugify(&annotation.comment));
            let x = clamp(annotation.rect.x * dpr - 12.0, 0.0, image_width as f64);
            let y = clamp(annotation.rect.y * dpr - 12.0, 0.0, image_height as f64);
            let width = clamp(annotation.rect.width * dpr + 24.0, 1.0, image_width as f64 - x);
            let height = clamp(annotation.rect.height * dpr + 24.0, 1.0, image_height as f64 - y);
            if width < 1.0 || height < 1.0 {
                continue;
            }
            let crop =
                image::imageops::crop_imm(&full, x as u32, y as u32, width as u32, height as u32)
                    .to_image();
            // 单条裁切失败不影响整体导出
            if crop
                .save_with_format(shots_dir.join(&name), image::ImageFormat::Png)
                .is_ok()
            {
                files.push(format!("screenshots/{name}"));
            }
        }
    }

    let exported_at = Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true);
    let viewport = data.annotations.first().and_then(viewport_of);
    let record = ExportRecord {
        round_id: &round_id,
        exported_at: &exported_at,
        view: &data.view,
        note: &data.note,
        viewport,
        dpr,
        app_version: &data.app_version,
        annotations: &data.annotations,
    };

    let json_body = json!({
        "roundId": round_id,
        "exportedAt": exported_at,
        "view": data.view,
        "note": data.note,
        "viewport": viewport.map(|(width, height)| json!({ "width": width, "height": height })),
        "devicePixelRatio": dpr,
        "appVersion": data.app_version,
        "annotations": raw_annotations
    });
    fs::write(
        dir.join("annotations.json"),
        serde_json::to_string_pretty(&json_body).map_err(|e| e.to_string())?,
    )
    .map_err(|e| format!("写入 annotations.json 失败: {e}"))?;
    files.push("annotations.json".to_string());

    fs::write(dir.join("annotations.md"), render_markdown(&record))
        .map_err(|e| format!("写入 annotations.md 失败: {e}"))?;
    files.push("annotations.md".to_string());

    if !dir.join("README.md").exists() {
        fs::write(dir.join("README.md"), render_round_readme(&record))
            .map_err(|e| format!("写入 README.md 失败: {e}"))?;
        files.push("README.md".to_string());
    }

    Ok(json!({
        "roundId": round_id,
        "roundDir": dir.to_string_lossy(),
        "files": files
    }))
}

fn count_rounds(rounds_dir: &Path) -> usize {
    fs::read_dir(rounds_dir)
        .map(|entries| entries.flatten().filter(|e| e.path().is_dir()).count())
        .unwrap_or(0)
}

fn stamp() -> String {
    Local::now().format("%Y%m%d-%H%M").to_string()
}

fn viewport_of(annotation: &Annotation) -> Option<(f64, f64)> {
    let viewport = annotation.viewport.as_ref()?;
    Some((
        viewport.get("width")?.as_f64()?,
        viewport.get("height")?.as_f64()?,
    ))
}

/// 抓整页：等界面重绘（旧版 invalidate + 320ms），再抓客户区像素
fn capture_full(app: &tauri::AppHandle) -> Option<image::RgbaImage> {
    let win = app.get_webview_window("main")?;
    std::thread::sleep(Duration::from_millis(320));
    let (width, height, rgba) = shot::capture_rgba(&win).ok()?;
    image::RgbaImage::from_raw(width, height, rgba)
}

/* -------------------------------------------------------------- Markdown */

fn render_round_readme(record: &ExportRecord) -> String {
    let viewport = match record.viewport {
        Some((width, height)) => format!("{width}×{height}"),
        None => "未知".to_string(),
    };
    let note = if record.note.is_empty() {
        "（待补充）".to_string()
    } else {
        record.note.to_string()
    };
    format!(
        "# 轮次 {round_id}

> 由页面标注工具自动生成，请在完成本轮改动后补全下面的小节。

## 背景

<!-- 这一轮要解决什么问题、由谁提出 -->

{note}

## 反馈标注

共 {count} 条，详见 [`annotations.md`](annotations.md) 与 [`annotations.json`](annotations.json)。

导出时视口 {viewport}，界面 {view}，应用版本 {app_version}。

## 改动

<!-- 逐条说明做了什么，以及为什么这么做（含取舍） -->

## 验证

<!-- 跑了哪些测试、截图证据、以及没能验证到的部分 -->

## 遗留

<!-- 本轮没做、留待下轮的事项 -->
",
        round_id = record.round_id,
        note = note,
        count = record.annotations.len(),
        viewport = viewport,
        view = record.view,
        app_version = record.app_version,
    )
}

fn render_markdown(record: &ExportRecord) -> String {
    let mut lines: Vec<String> = Vec::new();
    lines.push(format!("# 界面标注 · 轮次 {}", record.round_id));
    lines.push(String::new());
    lines.push(format!("- 导出时间：{}", record.exported_at));
    lines.push(format!("- 所在界面：{}", record.view));
    if let Some((width, height)) = record.viewport {
        lines.push(format!("- 视口：{width}×{height}（DPR {}）", record.dpr));
    }
    lines.push(format!("- 应用版本：{}", record.app_version));
    lines.push(format!("- 标注条数：{}", record.annotations.len()));
    if !record.note.is_empty() {
        lines.push(String::new());
        lines.push("## 总体说明".to_string());
        lines.push(String::new());
        lines.push(record.note.to_string());
    }
    lines.push(String::new());
    lines.push("## 逐条标注".to_string());
    lines.push(String::new());

    for annotation in record.annotations {
        lines.extend(render_one(annotation));
    }
    lines.join("\n")
}

fn render_one(annotation: &Annotation) -> Vec<String> {
    let target = &annotation.target;
    let mut lines: Vec<String> = Vec::new();
    let title = first_line(&annotation.comment);
    lines.push(format!(
        "### {}. {}",
        annotation.index,
        if title.is_empty() { "(无标题)" } else { &title }
    ));
    lines.push(String::new());
    lines.push(format!(
        "- **类别**：{} ｜ **优先级**：{} ｜ **方式**：{}",
        category_label(&annotation.category),
        severity_label(&annotation.severity),
        if annotation.kind == "element" { "点选元素" } else { "框选区域" }
    ));
    if let Some(component) = target.component.as_ref().filter(|value| !value.is_empty()) {
        lines.push(format!("- **组件**：`{component}`"));
    }
    lines.push(format!("- **选择器**：`{}`", target.selector));
    if !target.parent_chain.is_empty() {
        let chain = target
            .parent_chain
            .iter()
            .map(|item| format!("`{item}`"))
            .collect::<Vec<_>>()
            .join(" ← ");
        lines.push(format!("- **父级链**：{chain}"));
    }
    if !target.text.is_empty() {
        let collapsed = target.text.split_whitespace().collect::<Vec<_>>().join(" ");
        lines.push(format!("- **元素文本**：{}", truncate(&collapsed, 120)));
    }
    lines.push(format!(
        "- **位置尺寸**：x={} y={} {}×{}",
        annotation.rect.x.round() as i64,
        annotation.rect.y.round() as i64,
        annotation.rect.width.round() as i64,
        annotation.rect.height.round() as i64
    ));
    let styles = render_style_pairs(&target.styles);
    if !styles.is_empty() {
        lines.push(format!("- **关键样式**：{styles}"));
    }
    let attributes = render_attributes(&target.attributes);
    if !attributes.is_empty() {
        lines.push(format!("- **属性**：{attributes}"));
    }
    lines.push(String::new());
    lines.push(
        annotation
            .comment
            .split('\n')
            .map(|line| format!("> {line}"))
            .collect::<Vec<_>>()
            .join("\n"),
    );
    lines.push(String::new());
    lines
}

fn render_style_pairs(map: &Map<String, Json>) -> String {
    let mut items: Vec<(&String, &Json)> = map.iter().collect();
    items.sort_by(|a, b| a.0.cmp(b.0));
    items
        .iter()
        .map(|(key, value)| format!("{key}: {}", json_text(value)))
        .collect::<Vec<_>>()
        .join("; ")
}

fn render_attributes(map: &Map<String, Json>) -> String {
    let mut items: Vec<(&String, &Json)> = map.iter().collect();
    items.sort_by(|a, b| a.0.cmp(b.0));
    items
        .iter()
        .map(|(key, value)| format!("{key}=\"{}\"", truncate(&json_text(value), 40)))
        .collect::<Vec<_>>()
        .join(" ")
}

fn json_text(value: &Json) -> String {
    match value {
        Json::String(text) => text.clone(),
        other => other.to_string(),
    }
}

fn category_label(category: &str) -> &'static str {
    match category {
        "style" => "视觉样式",
        "layout" => "布局结构",
        "content" => "文案内容",
        "behavior" => "交互行为",
        "bug" => "缺陷",
        _ => "其它",
    }
}

fn severity_label(severity: &str) -> &'static str {
    match severity {
        "must" => "必须改",
        "should" => "建议改",
        _ => "锦上添花",
    }
}

fn first_line(text: &str) -> String {
    text.lines().next().unwrap_or("").trim().chars().take(60).collect()
}

fn truncate(text: &str, max: usize) -> String {
    if text.chars().count() <= max {
        return text.to_string();
    }
    let head: String = text.chars().take(max).collect();
    format!("{head}…")
}

/// 文件名片段：清掉半角与全角标点，中文保留
fn slugify(text: &str) -> String {
    const PUNCTUATION: &str = "/\\:*?\"<>|：；？！，。、（）【】《》“”‘’…—";
    let cleaned: String = first_line(text)
        .chars()
        .filter(|c| !PUNCTUATION.contains(*c))
        .collect();
    let mut slug = cleaned.split_whitespace().collect::<Vec<_>>().join("-");
    while slug.contains("--") {
        slug = slug.replace("--", "-");
    }
    let slug = slug.trim_matches('-').chars().take(28).collect::<String>();
    if slug.is_empty() {
        "annotation".to_string()
    } else {
        slug
    }
}

fn clamp(value: f64, min: f64, max: f64) -> f64 {
    value.max(min).min(max)
}
